File: quan.lua/src/test/message/user/UserInfo.lua
---
---用户信息
---代码自动生成，请勿手动修改
---

local _CodedBuffer = require("quan.message.CodedBuffer")
local _Message = require("quan.message.Message")
local test_message_role_RoleInfo = require("test.message.role.RoleInfo")
local RoleInfo = require("test.message.user.RoleInfo")

---
---用户信息
---
local UserInfo = {
    ---类名
    class = "test.message.user.UserInfo",
}

local function onSet(self, key, value)
    assert(not UserInfo[key], "不允许修改只读属性:" .. key)
    rawset(self, key, value)
end

local function toString(self)
    return "UserInfo{" ..
            "id=" .. tostring(self.id) ..
            ",name='" .. tostring(self.name) .. '\'' ..
            ",level=" .. tostring(self.level) ..
            ",type=" .. tostring(self.type) ..
            ",roleInfo1=" .. tostring(self.roleInfo1) ..
            ",roleInfo2=" .. tostring(self.roleInfo2) ..
            ",roleInfo3=" .. tostring(self.roleInfo3) ..
            ",roleList=" .. table.toString(self.roleList) ..
            ",roleSet=" .. table.toString(self.roleSet) ..
            ",roleMap=" .. table.toString(self.roleMap) ..
            ",f11=" .. tostring(self.f11) ..
            ",f12=" .. tostring(self.f12) ..
            ",f13=" .. tostring(self.f13) ..
            ",f14=" .. tostring(self.f14) ..
            ",f15=" .. tostring(self.f15) ..
            ",f16=" .. tostring(self.f16) ..
            ",f17=" .. tostring(self.f17) ..
            ",f18=" .. tostring(self.f18) ..
            ",alias='" .. tostring(self.alias) .. '\'' ..
            '}';
end

---元表
local meta = { __index = UserInfo, __newindex = onSet, __tostring = toString }

---
---[用户信息].构造
---@param args 参数列表可以为空
---
function UserInfo.new(args)
    assert(args == nil or type(args) == "table", "参数错误")
    args = args or {}

    local instance = {
        ---ID
        id = args.id or 0,
        ---名字
        name = args.name or "",
        ---等级
        level = args.level or 0,
        ---类型
        type = args.type or 0,
        ---角色信息
        roleInfo1 = args.roleInfo1,
        ---角色信息2
        roleInfo2 = args.roleInfo2 or RoleInfo.new(),
        ---角色信息2
        roleInfo3 = args.roleInfo3 or RoleInfo.new(),
        ---角色信息List
        roleList = args.roleList or {},
        ---角色信息Set
        roleSet = args.roleSet or {},
        ---角色信息Map
        roleMap = args.roleMap or {},
        f11 = args.f11 or "",
        f12 = args.f12 or false,
        f13 = args.f13 or false,
        f14 = args.f14 or 0,
        f15 = args.f15 or 0.0,
        f16 = args.f16 or 0.0,
        f17 = args.f17 or 0.0,
        f18 = args.f18 or 0.0,
        alias = args.alias,
    }

    instance = setmetatable(instance, meta)
    return instance
end

setmetatable(UserInfo, { __call = UserInfo.new })

---
---[用户信息].编码
---@param buffer quan.message.CodedBuffer 可以为空
---@return quan.message.CodedBuffer
---
function UserInfo:encode(buffer)
    assert(type(self) == "table" and self.class == UserInfo.class, "参数[self]类型错误")
    assert(buffer == nil or type(buffer) == "table" and buffer.class == _CodedBuffer.class, "参数[buffer]类型错误")

    buffer = buffer or _CodedBuffer.new()

    if self.id ~= 0 then
        _Message.writeTag(buffer, 4)
        buffer:writeInt(self.id)
    end

    if #self.name > 0 then
        _Message.writeTag(buffer, 11)
        buffer:writeString(self.name)
    end

    if self.level ~= 0 then
        _Message.writeTag(buffer, 12)
        buffer:writeInt(self.level)
    end

    if self.type ~= nil and self.type ~= 0 then
        _Message.writeTag(buffer, 16)
        buffer:writeInt(self.type)
    end

    if self.roleInfo1 ~= nil then
        _Message.writeTag(buffer, 23)
        local roleInfo1Buffer = _CodedBuffer.new()
        test_message_role_RoleInfo.encode(self.roleInfo1, roleInfo1Buffer)
        buffer:writeBuffer(roleInfo1Buffer)
    end

    _Message.writeTag(buffer, 27)
    local roleInfo2Buffer = _CodedBuffer.new()
    RoleInfo.encode(self.roleInfo2, roleInfo2Buffer)
    buffer:writeBuffer(roleInfo2Buffer)

    _Message.writeTag(buffer, 31)
    local roleInfo3Buffer = _CodedBuffer.new()
    RoleInfo.encode(self.roleInfo3, roleInfo3Buffer)
    buffer:writeBuffer(roleInfo3Buffer)

    if #self.roleList > 0 then
        _Message.writeTag(buffer, 35)
        local roleListBuffer = _CodedBuffer.new()
        roleListBuffer:writeInt(#self.roleList)
        for i, value in ipairs(self.roleList) do
            test_message_role_RoleInfo.encode(value, roleListBuffer)
        end
        buffer:writeBuffer(roleListBuffer)
    end

    if #self.roleSet > 0 then
        _Message.writeTag(buffer, 39)
        local roleSetBuffer = _CodedBuffer.new()
        roleSetBuffer:writeInt(#self.roleSet)
        for i, value in ipairs(self.roleSet) do
            test_message_role_RoleInfo.encode(value, roleSetBuffer)
        end
        buffer:writeBuffer(roleSetBuffer)
    end

    local roleMapSize = table.size(self.roleMap)
    if roleMapSize > 0 then
        _Message.writeTag(buffer, 43)
        local roleMapBuffer = _CodedBuffer.new()
        roleMapBuffer:writeInt(roleMapSize)
        for key, value in pairs(self.roleMap) do
            roleMapBuffer:writeInt(key)
            test_message_role_RoleInfo.encode(value, roleMapBuffer)
        end
        buffer:writeBuffer(roleMapBuffer)
    end

    if #self.f11 > 0 then
        _Message.writeTag(buffer, 47)
        buffer:writeBytes(self.f11)
    end

    if self.f12 then
        _Message.writeTag(buffer, 48)
        buffer:writeBool(self.f12)
    end

    if self.f13 then
        _Message.writeTag(buffer, 52)
        buffer:writeBool(self.f13)
    end

    if self.f14 ~= 0 then
        _Message.writeTag(buffer, 56)
        buffer:writeShort(self.f14)
    end

    if self.f15 ~= 0 then
        _Message.writeTag(buffer, 61)
        buffer:writeFloat(self.f15)
    end

    if self.f16 ~= 0 then
        _Message.writeTag(buffer, 64)
        buffer:writeFloat(self.f16, 2)
    end

    if self.f17 ~= 0 then
        _Message.writeTag(buffer, 70)
        buffer:writeDouble(self.f17)
    end

    if self.f18 ~= 0 then
        _Message.writeTag(buffer, 72)
        buffer:writeDouble(self.f18, 2)
    end

    if self.alias ~= nil then
        _Message.writeTag(buffer, 79)
        buffer:writeString(self.alias)
    end

    _Message.writeTag(buffer, 0);

    return buffer
end

---
---[用户信息].解码
---@param buffer quan.message.CodedBuffer 不能为空
---@param self test.message.user.UserInfo 可以为空
---@return test.message.user.UserInfo
---
function UserInfo.decode(buffer, self)
    assert(type(buffer) == "table" and buffer.class == _CodedBuffer.class, "参数[buffer]类型错误")
    assert(self == nil or type(self) == "table" and self.class == UserInfo.class, "参数[self]类型错误")

    self = self or UserInfo.new()

    while true do
        local tag = _Message.readTag(buffer)
        if tag == 0 then
            break
        elseif tag == 4 then
            self.id = buffer:readInt()
        elseif tag == 11 then
            self.name = buffer:readString()
        elseif tag == 12 then
            self.level = buffer:readInt()
        elseif tag == 16 then
            self.type = buffer:readInt()
        elseif tag == 23 then
            buffer:readInt()
            self.roleInfo1 = test_message_role_RoleInfo.decode(buffer)
        elseif tag == 27 then
            buffer:readInt()
            self.roleInfo2 = RoleInfo.decode(buffer, self.roleInfo2)
        elseif tag == 31 then
            buffer:readInt()
            self.roleInfo3 = RoleInfo.decode(buffer, self.roleInfo3)
        elseif tag == 35 then
            buffer:readInt()
            for i = 1, buffer:readInt() do
                self.roleList[i] = test_message_role_RoleInfo.decode(buffer)
            end
        elseif tag == 39 then
            buffer:readInt()
            for i = 1, buffer:readInt() do
                self.roleSet[i] = test_message_role_RoleInfo.decode(buffer)
            end
        elseif tag == 43 then
            buffer:readInt()
            for i = 1, buffer:readInt() do
                self.roleMap[buffer:readInt()] = test_message_role_RoleInfo.decode(buffer)
            end
        elseif tag == 47 then
            self.f11 = buffer:readBytes()
        elseif tag == 48 then
            self.f12 = buffer:readBool()
        elseif tag == 52 then
            self.f13 = buffer:readBool()
        elseif tag == 56 then
            self.f14 = buffer:readShort()
        elseif tag == 61 then
            self.f15 = buffer:readFloat()
        elseif tag == 64 then
            self.f16 = buffer:readFloat(2)
        elseif tag == 70 then
            self.f17 = buffer:readDouble()
        elseif tag == 72 then
            self.f18 = buffer:readDouble(2)
        elseif tag == 79 then
            self.alias = buffer:readString()
        else
            _Message.skipField(tag, buffer)
        end
    end

    return self
end

UserInfo = table.readOnly(UserInfo)
return UserInfo

File: quan.cs/Test/Message/User/UserInfo.cs
using Quan.Message;
using Quan.Utils;
using System;
using System.Collections.Generic;

namespace Test.Message.User
{
    /// <summary>
    /// 用户信息<br/>
    /// 代码自动生成，请勿手动修改
    /// </summary>
    public class UserInfo : Bean
    {
        /// <summary>
        /// ID
        /// </summary>
        public int id { get; set; }

        private string _name = "";

        /// <summary>
        /// 名字
        /// </summary>
        public string name
        {
            get => _name;
            set => _name = value ?? throw new NullReferenceException();
        }

        /// <summary>
        /// 等级
        /// </summary>
        public int level { get; set; }

        /// <summary>
        /// 类型
        /// </summary>
        public UserType type { get; set; }

        /// <summary>
        /// 角色信息
        /// </summary>
        public Test.Message.Role.RoleInfo roleInfo1 { get; set; }

        private RoleInfo _roleInfo2 = new RoleInfo();

        /// <summary>
        /// 角色信息2
        /// </summary>
        public RoleInfo roleInfo2
        {
            get => _roleInfo2;
            set => _roleInfo2 = value ?? throw new NullReferenceException();
        }

        private RoleInfo _roleInfo3 = new RoleInfo();

        /// <summary>
        /// 角色信息2
        /// </summary>
        public RoleInfo roleInfo3
        {
            get => _roleInfo3;
            set => _roleInfo3 = value ?? throw new NullReferenceException();
        }

        /// <summary>
        /// 角色信息List
        /// </summary>
        public IList<Test.Message.Role.RoleInfo> roleList { get; } = new List<Test.Message.Role.RoleInfo>();

        /// <summary>
        /// 角色信息Set
        /// </summary>
        public ISet<Test.Message.Role.RoleInfo> roleSet { get; } = new HashSet<Test.Message.Role.RoleInfo>();

        /// <summary>
        /// 角色信息Map
        /// </summary>
        public IDictionary<int, Test.Message.Role.RoleInfo> roleMap { get; } = new Dictionary<int, Test.Message.Role.RoleInfo>();

        private byte[] _f11 = Array.Empty<byte>();

        public byte[] f11
        {
            get => _f11;
            set => _f11 = value ?? throw new NullReferenceException();
        }

        public bool f12 { get; set; }

        public bool f13 { get; set; }

        public short f14 { get; set; }

        public float f15 { get; set; }

        private float _f16;

        public float f16
        {
            get => _f16;
            set
            {
                CodedBuffer.ValidateScale(value, 2);
                _f16 = value;
            }
        }

        public double f17 { get; set; }

        private double _f18;

        public double f18
        {
            get => _f18;
            set
            {
                CodedBuffer.ValidateScale(value, 2);
                _f18 = value;
            }
        }

        public string alias { get; set; }


        public override void Encode(CodedBuffer buffer)
        {
            base.Encode(buffer);

            Validate();

            if (id != 0)
            {
                WriteTag(buffer, 4);
                buffer.WriteInt(id);
            }

            if (name.Length > 0)
            {
                WriteTag(buffer, 11);
                buffer.WriteString(name);
            }

            if (level != 0)
            {
                WriteTag(buffer, 12);
                buffer.WriteInt(level);
            }

            if (type != 0)
            {
                WriteTag(buffer, 16);
                buffer.WriteInt((int) type);
            }

            if (roleInfo1 != null)
            {
                WriteTag(buffer, 23);
                var roleInfo1Buffer = new CodedBuffer();
                roleInfo1.Encode(roleInfo1Buffer);
                buffer.WriteBuffer(roleInfo1Buffer);
            }

            WriteTag(buffer, 27);
            var roleInfo2Buffer = new CodedBuffer();
            roleInfo2.Encode(roleInfo2Buffer);
            buffer.WriteBuffer(roleInfo2Buffer);

            WriteTag(buffer, 31);
            var roleInfo3Buffer = new CodedBuffer();
            roleInfo3.Encode(roleInfo3Buffer);
            buffer.WriteBuffer(roleInfo3Buffer);

            if (roleList.Count > 0)
            {
                WriteTag(buffer, 35);
                var roleListBuffer = new CodedBuffer();
                roleListBuffer.WriteInt(roleList.Count);
                foreach (var roleListValue in roleList)
                {
                    roleListValue.Encode(roleListBuffer);
                }
                buffer.WriteBuffer(roleListBuffer);
            }

            if (roleSet.Count > 0)
            {
                WriteTag(buffer, 39);
                var roleSetBuffer = new CodedBuffer();
                roleSetBuffer.WriteInt(roleSet.Count);
                foreach (var roleSetValue in roleSet)
                {
                    roleSetValue.Encode(roleSetBuffer);
                }
                buffer.WriteBuffer(roleSetBuffer);
            }

            if (roleMap.Count > 0)
            {
                WriteTag(buffer, 43);
                var roleMapBuffer = new CodedBuffer();
                roleMapBuffer.WriteInt(roleMap.Count);
                foreach (var roleMapKey in roleMap.Keys)
                {
                    roleMapBuffer.WriteInt(roleMapKey);
                    roleMap[roleMapKey].Encode(roleMapBuffer);
                }
                buffer.WriteBuffer(roleMapBuffer);
            }

            if (f11.Length > 0)
            {
                WriteTag(buffer, 47);
                buffer.WriteBytes(f11);
            }

            if (f12)
            {
                WriteTag(buffer, 48);
                buffer.WriteBool(f12);
            }

            if (f13)
            {
                WriteTag(buffer, 52);
                buffer.WriteBool(f13);
            }

            if (f14 != 0)
            {
                WriteTag(buffer, 56);
                buffer.WriteShort(f14);
            }

            if (f15 != 0)
            {
                WriteTag(buffer, 61);
                buffer.WriteFloat(f15);
            }

            if (f16 != 0)
            {
                WriteTag(buffer, 64);
                buffer.WriteFloat(f16, 2);
            }

            if (f17 != 0)
            {
                WriteTag(buffer, 70);
                buffer.WriteDouble(f17);
            }

            if (f18 != 0)
            {
                WriteTag(buffer, 72);
                buffer.WriteDouble(f18, 2);
            }

            if (alias != null)
            {
                WriteTag(buffer, 79);
                buffer.WriteString(alias);
            }

            WriteTag(buffer,0);
        }

        public override void Decode(CodedBuffer buffer)
        {
            base.Decode(buffer);

            for (var tag = ReadTag(buffer); tag != 0; tag = ReadTag(buffer))
            {
                switch (tag)
                {
                    case 4:
                        id = buffer.ReadInt();
                        break;
                    case 11:
                        name = buffer.ReadString();
                        break;
                    case 12:
                        level = buffer.ReadInt();
                        break;
                    case 16:
                        type = (UserType) buffer.ReadInt();
                        break;
                    case 23:
                        buffer.ReadInt();
                        roleInfo1 = roleInfo1 ?? new Test.Message.Role.RoleInfo();
                        roleInfo1.Decode(buffer);
                        break;
                    case 27:
                        buffer.ReadInt();
                        roleInfo2.Decode(buffer);
                        break;
                    case 31:
                        buffer.ReadInt();
                        roleInfo3.Decode(buffer);
                        break;
                    case 35:
                        buffer.ReadInt();
                        var roleList_Size = buffer.ReadInt();
                        for (var roleList_i = 0; roleList_i < roleList_Size; roleList_i++) 
                        {
                            var roleListValue = new Test.Message.Role.RoleInfo();
                            roleListValue.Decode(buffer);
                            roleList.Add(roleListValue);
                        }
                        break;
                    case 39:
                        buffer.ReadInt();
                        var roleSet_Size = buffer.ReadInt();
                        for (var roleSet_i = 0; roleSet_i < roleSet_Size; roleSet_i++) 
                        {
                            var roleSetValue = new Test.Message.Role.RoleInfo();
                            roleSetValue.Decode(buffer);
                            roleSet.Add(roleSetValue);
                        }
                        break;
                    case 43:
                        buffer.ReadInt();
                        var roleMap_Size = buffer.ReadInt();
                        for (var roleMap_i = 0; roleMap_i < roleMap_Size; roleMap_i++)
                        {
                            var roleMapKey = buffer.ReadInt();
                            var roleMapValue = new Test.Message.Role.RoleInfo();
                            roleMapValue.Decode(buffer);
                            roleMap.Add(roleMapKey, roleMapValue);
                        }
                        break;
                    case 47:
                        f11 = buffer.ReadBytes();
                        break;
                    case 48:
                        f12 = buffer.ReadBool();
                        break;
                    case 52:
                        f13 = buffer.ReadBool();
                        break;
                    case 56:
                        f14 = buffer.ReadShort();
                        break;
                    case 61:
                        f15 = buffer.ReadFloat();
                        break;
                    case 64:
                        f16 = buffer.ReadFloat(2);
                        break;
                    case 70:
                        f17 = buffer.ReadDouble();
                        break;
                    case 72:
                        f18 = buffer.ReadDouble(2);
                        break;
                    case 79:
                        alias = buffer.ReadString();
                        break;
                    default:
                        SkipField(tag, buffer);
                        break;
                }
            }

            Validate();
        }

        public override void Validate()
        {
            base.Validate();

            ValidateNull(name, "字段[name]");
            ValidateNull(roleInfo2, "字段[roleInfo2]");
            ValidateNull(roleInfo3, "字段[roleInfo3]");
            ValidateNull(f11, "字段[f11]");
            CodedBuffer.ValidateScale(f16, 2, "字段[f16]");
            CodedBuffer.ValidateScale(f18, 2, "字段[f18]");
        }

        public override string ToString()
        {
            return "UserInfo{" +
                   "id=" + id.ToString2() +
                   ",name='" + name + '\'' +
                   ",level=" + level.ToString2() +
                   ",type=" + type.ToString2() +
                   ",roleInfo1=" + roleInfo1.ToString2() +
                   ",roleInfo2=" + roleInfo2.ToString2() +
                   ",roleInfo3=" + roleInfo3.ToString2() +
                   ",roleList=" + roleList.ToString2() +
                   ",roleSet=" + roleSet.ToString2() +
                   ",roleMap=" + roleMap.ToString2() +
                   ",f11=" + f11.ToString2() +
                   ",f12=" + f12.ToString2() +
                   ",f13=" + f13.ToString2() +
                   ",f14=" + f14.ToString2() +
                   ",f15=" + f15.ToString2() +
                   ",f16=" + f16.ToString2() +
                   ",f17=" + f17.ToString2() +
                   ",f18=" + f18.ToString2() +
                   ",alias='" + alias + '\'' +
                   '}';
        }
    }
}